src/range.rs

```diff
@@ -308,6 +308,100 @@ impl<V: Ord> Range<V> {
     }
 }
 
+/// Implementing `PartialOrd` for start `Bound` of an interval.
+///
+/// ```text
+/// left:   ∞-------]
+/// right:    [-----]
+/// left is smaller, since it starts earlier.
+///
+/// left:   [-----]
+/// right:  ]-----]
+/// left is smaller, since it starts earlier.
+/// ```
+fn cmp_bounds_start<V: PartialOrd>(left: Bound<&V>, right: Bound<&V>) -> Option<Ordering> {
+    match (left, right) {
+        (Unbounded, Unbounded) => Some(Ordering::Equal),
+        (Included(_left), Unbounded) => Some(Ordering::Greater),
+        (Excluded(_left), Unbounded) => Some(Ordering::Greater),
+        (Unbounded, Included(_right)) => Some(Ordering::Less),
+        (Included(left), Included(right)) => left.partial_cmp(right),
+        (Excluded(left), Included(right)) => match left.partial_cmp(right)? {
+            Ordering::Less => Some(Ordering::Less),
+            Ordering::Equal => Some(Ordering::Greater),
+            Ordering::Greater => Some(Ordering::Greater),
+        },
+        (Unbounded, Excluded(_right)) => Some(Ordering::Less),
+        (Included(left), Excluded(right)) => match left.partial_cmp(right)? {
+            Ordering::Less => Some(Ordering::Less),
+            Ordering::Equal => Some(Ordering::Less),
+            Ordering::Greater => Some(Ordering::Greater),
+        },
+        (Excluded(left), Excluded(right)) => left.partial_cmp(right),
+    }
+}
+
+/// Implementing `PartialOrd` for end `Bound` of an interval.
+///
+/// We flip the unbounded ranges from `-∞` to `∞`, while `V`-valued bounds checks remain the same.
+///
+/// ```text
+/// left:   [--------inf
+/// right:  [-----]
+/// left is greater, since it starts earlier.
+///
+/// left:   [-----[
+/// right:  [-----]
+/// left is smaller, since it ends earlier.
+/// ```
+fn cmp_bounds_end<V: PartialOrd>(left: Bound<&V>, right: Bound<&V>) -> Option<Ordering> {
+    match (left, right) {
+        (Unbounded, Unbounded) => Some(Ordering::Equal),
+        (Included(_left), Unbounded) => Some(Ordering::Less),
+        (Excluded(_left), Unbounded) => Some(Ordering::Less),
+        (Unbounded, Included(_right)) => Some(Ordering::Greater),
+        (Included(left), Included(right)) => left.partial_cmp(right),
+        (Excluded(left), Included(right)) => match left.partial_cmp(right)? {
+            Ordering::Less => Some(Ordering::Less),
+            Ordering::Equal => Some(Ordering::Greater),
+            Ordering::Greater => Some(Ordering::Greater),
+        },
+        (Unbounded, Excluded(_right)) => Some(Ordering::Greater),
+        (Included(left), Excluded(right)) => match left.partial_cmp(right)? {
+            Ordering::Less => Some(Ordering::Less),
+            Ordering::Equal => Some(Ordering::Less),
+            Ordering::Greater => Some(Ordering::Greater),
+        },
+        (Excluded(left), Excluded(right)) => left.partial_cmp(right),
+    }
+}
```

Comment:
It looks like this could _almost_ be `cmp_bounds_start(left, right).map(|o| o.reverse())`, but the `Included, Included`, `Excluded, Included` and `Included, Excluded` cases I think would be different.